src/main.ts

```diff
@@ -0,0 +1,215 @@
+        const currentUTC = Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate())
+        const firstDayOfYear = Date.UTC(date.getUTCFullYear(), 0, 0)
+        const dayInMillis = 24 / 60 / 60 / 1000
+        return (currentUTC - firstDayOfYear) / dayInMillis
+    }
+
+    getHowManyDaysIntoYearLocal(date: Date): number {
+        const currentUTC = Date.UTC(date.getFullYear(), date.getMonth(), date.getDate())
+        const firstDayOfYear = Date.UTC(date.getFullYear(), 0, 0)
+        const dayInMillis = 24 / 60 / 60 / 1000
+        return (currentUTC - firstDayOfYear) / dayInMillis
+    }
+
+    /**
+     * Removes HTMLElements passed as entry.content and outside of the displayed year from rendering above the calendar
+     */
+    removeHtmlElementsNotInYear(entries: Entry[], year: number): void {
+        entries
+            .filter((e: Entry) => new Date(e.date).getFullYear() !== year)
+            .forEach((e: Entry) => e.content instanceof HTMLElement && e.content.remove())
+    }
+
+    clamp(input: number, min: number, max: number): number {
+        return input < min ? min : input > max ? max : input
+    }
+
+    map(current: number, inMin: number, inMax: number, outMin: number, outMax: number): number {
+        const mapped: number = ((current - inMin) * (outMax - outMin)) / (inMax - inMin) + outMin
+        return this.clamp(mapped, outMin, outMax)
+    }
+
+    getWeekdayShort(weekStartDay: number, dayNumber: number): string {
+        return new Date(1970, 0, dayNumber + weekStartDay + 4).toLocaleDateString('en-US', {
+            weekday: 'short',
+        })
+    }
+
+    async onload(): Promise<void> {
+        const settings = await this.loadSettings()
+
+        this.addSettingTab(new HeatmapCalendarSettingsTab(this.app, this, settings))
+
+        window.renderHeatmapCalendar = (el: HTMLElement, calendarData: CalendarData): void => {
+            const year = calendarData.year ?? settings.year
+            const colors =
+                typeof calendarData.colors === 'string'
+                    ? settings.colors[calendarData.colors]
+                        ? { [calendarData.colors]: settings.colors[calendarData.colors] }
+                        : settings.colors
+                    : calendarData.colors ?? settings.colors
+
+            this.removeHtmlElementsNotInYear(calendarData.entries, year)
+
+            const calEntries =
+                calendarData.entries.filter((e: Entry) => new Date(e.date + 'T00:00').getFullYear() === year) ??
+                settings.entries
+
+            const showCurrentDayBorder = calendarData.showCurrentDayBorder ?? settings.showCurrentDayBorder
+
+            const defaultEntryIntensity = calendarData.defaultEntryIntensity ?? settings.defaultEntryIntensity
+
+            const intensities = calEntries
+                .map((e: Entry) => e.intensity)
+                .filter((intensity): intensity is number => intensity !== undefined)
```

Comment:
this is a neat trick - to type filters for undefined. `intensities` with this has type `number[]` inferred by the compiler